### PR TITLE
Add janet_parser_* bindings

### PR DESCRIPTION
**Pull request by iacore**
I have used this in a project of mine, which you can [read the parser code here](https://codeberg.org/iacore/xn--8st83p/src/commit/d0ffd9d6d320e9b9a4fa7c9cb93e39986baa4ff0/src/Parser.zig).

It could use some proper tests though.

### CODE REVIEW

File: src/janet.zig [hunk 1]
```diff
@@ -2291,3 +2291,62 @@ test "function call" {
...
+
+pub const Parser = struct {
+    inner: c.JanetParser,
+
+    pub fn init() @This() {
+        var this: @This() = undefined;
+        c.janet_parser_init(&this.inner);
+        return this;
+    }
+    pub fn deinit(this: *@This()) void {
```
Comment (greenfork) [inline src/janet.zig]: Can I ask to change this and all other cases to `self: *Parser`? All cases of `this` should probably be converted to `self`. We already use this style in this codebase and iirc Loris Cro did some evangelism on the topic that "`@This()` should be used in metaprogramming when you create types and you don't have the type's name".

File: src/janet.zig [hunk 2]
```diff
@@ -2291,3 +2291,62 @@ test "function call" {
...
+    }
+    pub fn flush(this: *@This()) void {
+        return c.janet_parser_flush(&this.inner);
+    }
+    pub fn eof(this: *@This()) void {
+        return c.janet_parser_eof(&this.inner);
+    }
+    pub fn hasMore(this: *@This()) c_int {
+        return c.janet_parser_has_more(&this.inner);
+    }
```
Comment (greenfork) [inline src/janet.zig]: I think this function should return a boolean

File: src/janet.zig [hunk 3]
```diff
@@ -2291,3 +2291,62 @@ test "function call" {
...
+    pub fn produce(this: *@This()) Janet {
+        return Janet.fromC(c.janet_parser_produce(&this.inner));
+    }
+    pub fn produceWrapped(this: *@This()) Janet {
+        return Janet.fromC(c.janet_parser_produce_wrapped(&this.inner));
+    }
+    pub fn @"error"(this: *@This()) [*c]const u8 {
+        return c.janet_parser_error(&this.inner);
+    }
```
Comment (greenfork) [inline src/janet.zig]: Looks like we know that the return value is an array, so we can write it like `[*]const u8`. And we also know that this a C string that ends with a null byte, so we can write it like so:

```suggestion
    pub fn @"error"(this: *@This()) [*:0]const u8 {
        return c.janet_parser_error(&this.inner);
    }
```

There will probably be more type casting necessary

**Review by greenfork — COMMENTED** [review]
Looks great, there are only some minor fixes!

File: src/janet.zig (hunk 2 above)
Comment (iacore) [reply src/janet.zig]: I just checked Janet's source code and it is a boolean...

File: src/janet.zig (hunk 1 above)
Comment (iacore) [reply src/janet.zig]: Sorry about the style. That was my usual style of programming. I don't have to remember the worked-on type name if I use `@This()` everywhere. Names are irrelevant to the code's meaning anyway.

**Comment by iacore** [comment]
I have applied your suggestions.

I am not sure whether I should use `std.mem.span` to return `[:0]const u8` in `Parser.@"error"`.

**Review by greenfork — APPROVED** [review]
Thank you, looks great!

File: src/janet.zig (hunk 1 above)
Comment (greenfork) [reply src/janet.zig]: Sure, it's just a style :^)

**Comment by greenfork** [comment]
> I have applied your suggestions.
> 
> I am not sure whether I should use `std.mem.span` to return `[:0]const u8` in `Parser.@"error"`.

I don't know about that too. If that works, then it's fine

**Comment by greenfork** [comment]
@iacore for your information, I also tagged a new release and updated the zon hash and url in the readme

**Pull request merged**